### ISSUE TITLE
All threads locked to single CPU core at a time

### ISSUE THREAD

**Issue by scroix** [oaubert/python-vlc#20]
Hi Oliver,

Firstly - great project. I've been investing a lot of effort into using these bindings.

It'd been fairly smooth sailing until I started playing back some heavier content (h264@2160p). I noticed that there was a considerable performance difference between native VLC and the python bindings.

It turns out that the **decoding threads appear limited to one CPU core** when ran as part of the bindings. Below are some snippets of the process threads on the two players, _native_ and through the _python bindings_. Note: I'm running on Windows, but will soon attempt the same on Linux.

![eg_nativethreads](https://cloud.githubusercontent.com/assets/9277107/20639342/9b986fce-b414-11e6-8d62-1e479c419b2e.png)

![eg_pythonthreads](https://cloud.githubusercontent.com/assets/9277107/20639343/9ec7d3a6-b414-11e6-8ee0-2831ebe93fb2.png)

In the first example, you'll see the ffmpeg decoders are spread quite neatly over my four cores. They're utilising roughly 45% of total CPU usage. In the second example, all the heavy lifting is being done by one maxed out core, for a total of 25% total CPU usage.

I've tried to do my homework before creating an issue, and suspect that this might be a product of Python's GIL - but I wasn't sure.

I've attempted a couple of things:

- Utilising different versions of Python.
- Running the bindings in their purest form.
`vlc.py [content]`
- Emulating the affinity of native VLC to emulate this behaviour - it replicates the results exactly.
- Forcing multiple threads as a passed parameter.
`--avcodec-threads=`

I'm stuck at this point - is this a limitation of the bindings?

Appreciate your time,

Thank you!

**Comment by oaubert** [comment]
Hi. I do not think python is in cause here, since it merely wraps the libvlc API, which should be responsible for launching threads appropriately. One way to more precisely pinpoint the issue would be to write a simple C program to play a movie through libvlc and check if you see the same symptoms.

**Comment by scroix** [comment]
Thanks for the quick response.

Good point about VLC wrapping the libvlc API, not actually launching the threads itself.

I went ahead and ran the same tests on Ubuntu 16.4.1 to see if the same issues occurred - surprise! It appears to be working on Linux fine.

**Native VLC Forced to 1 Core (for comparison)**
<img width="725" alt="native_vlc_force1" src="https://cloud.githubusercontent.com/assets/9277107/20639678/62854ba8-b41e-11e6-8a92-1d9fb810a0a9.png">

**Python Bindings**
<img width="724" alt="vlc_bindings" src="https://cloud.githubusercontent.com/assets/9277107/20639679/64c7c2f6-b41e-11e6-8710-63d362f3dfb0.png">

I'm still keen to see what the issue is on Windows, I'll try write that program you suggested.

Cheers.

**Comment by scroix** [comment]
I didn't end up writing that C program - but, I did notice some interesting things, and have found a resolution.

In addition to working fine on Linux (as above), HEVC decoding was fine, and more importantly I tracked down the issue as occurring only in recent versions of VLC - 2.1.5 - 2.2.5. Versions before and after (the nightly 3.0s) appear to be fine.

Bizarre! 

But regardless, thank you for your help. I'm just going to avoid those versions for now. I suspect, as you suggested, the bindings aren't part of the problem.
